zebra-state/src/service/check/utxo.rs

```diff
@@ -60,52 +57,150 @@ pub fn transparent_double_spends(
+            // which can be rejected by later contextual checks.
+            // This is a particular issue for v5 transactions,
+            // because their authorizing data is only bound to the block data
+            // during contextual validation (#2336).
             //
-            // because we are in the non-finalized state, we need to check spends within the same block,
-            // spent non-finalized UTXOs, and unspent non-finalized and finalized UTXOs.
-
-            if let Some(output) = prepared.new_outputs.get(spend) {
-                // reject the spend if it uses an output from this block,
-                // but the output was not created by an earlier transaction
-                //
-                // we know the spend is invalid, because transaction IDs are unique
-                //
-                // (transaction IDs also commit to transaction inputs,
-                // so it should be cryptographically impossible for a transaction
-                // to spend its own outputs)
-                if output.tx_index_in_block >= spend_tx_index_in_block {
-                    return Err(EarlyTransparentSpend { outpoint: *spend });
-                } else {
-                    // a unique spend of a previous transaction's output is ok
-                    continue;
-                }
-            }
+            // We don't want to use UTXOs from invalid pending blocks,
+            // so we check transparent coinbase maturity and shielding
+            // using known valid UTXOs during non-finalized chain validation.
 
-            if non_finalized_chain_spent_utxos.contains(spend) {
-                // reject the spend if its UTXO is already spent in the
-                // non-finalized parent chain
-                return Err(DuplicateTransparentSpend {
-                    outpoint: *spend,
-                    location: "the non-finalized chain",
-                });
-            }
-
-            if !non_finalized_chain_unspent_utxos.contains_key(spend)
-                && finalized_state.utxo(spend).is_none()
-            {
-                // we don't keep spent UTXOs in the finalized state,
-                // so all we can say is that it's missing from both
-                // the finalized and non-finalized chains
-                // (it might have been spent in the finalized state,
-                // or it might never have existed in this chain)
-                return Err(MissingTransparentOutput {
-                    outpoint: *spend,
-                    location: "the non-finalized and finalized chain",
-                });
+            let spend_restriction = transaction.coinbase_spend_restriction(prepared.height);
+            if cfg!(not(test)) {
+                // TODO: fix proptests to produce valid UTXO spends (#ticket TODO)
+                //       and unconditionally enable this check
+                transparent_coinbase_spend(*spend, spend_restriction, utxo)?;
             }
         }
```

Comment:
@oxarbitrage at this point we know we have the correct UTXOs, so we can validate `remaining_transaction_value` here.